### PR TITLE
enhancement(file source): Add optional byte offset to file source events

### PR DESCRIPTION
**Pull request by sproberts92**
Motivation: If an application produces log lines faster than the resolution of the timestamp used additional information is needed to maintain ordering, and form part of a unique document id used in downstream services such as Elasticsearch.

Off by default, turned on by setting a value in config for `offset_key`.

This addresses issue #6633.

### CODE REVIEW

**Comment by sproberts92** [comment]
Had an off-by-one issue with the offset passed in the test, the confusion came from the `line` variable having the `newline` stripped, but the offset of course still counting it. Makes perfect sense in a real file, but a little less obvious in a test where the line is being created synthetically.  Have expanded the comment to make it clearer why the value is one higher. 

To demonstrate how it looks when reading from a real file, I added a `println` to `create_event`:
```
$ cat test_log
0123
0123
0123
0123
0123

$ ./target/debug/vector -c config/vector.toml
line_start_offset=0 line_end_offset=5 line_len=4
line_start_offset=5 line_end_offset=10 line_len=4
line_start_offset=10 line_end_offset=15 line_len=4
line_start_offset=15 line_end_offset=20 line_len=4
line_start_offset=20 line_end_offset=25 line_len=4
{"file":"test_log","host":"foo","message":"0123","offset":"0","source_type":"file","
timestamp":"2022-07-06T10:50:06.440885177Z"}
{"file":"test_log","host":"foo","message":"0123","offset":"5","source_type":"file","
timestamp":"2022-07-06T10:50:06.441392827Z"}
{"file":"test_log","host":"foo","message":"0123","offset":"10","source_type":"file",
"timestamp":"2022-07-06T10:50:06.441807939Z"}
{"file":"test_log","host":"foo","message":"0123","offset":"15","source_type":"file",
"timestamp":"2022-07-06T10:50:06.442170874Z"}
{"file":"test_log","host":"foo","message":"0123","offset":"20","source_type":"file",
"timestamp":"2022-07-06T10:50:06.442540067Z"}
```
Each line has 4 characters, plus a newline. We see that `line_len` is 4, so  without the newline, and the offset is 5 as expected.

**Comment by atibdialpad** [comment]
@jszwedko Ping !

**Comment by jszwedko** [comment]
> @jszwedko Ping !

Apologies for the delay! We'll get this reviewed next week. The changes look straight-forward enough.

File: src/sources/file.rs
Comment (bruceg) [inline src/sources/file.rs:552]: This calling convention is getting unwieldy, particularly with the interleaving of the source fields and the output key names. I'm actually surprised clippy didn't complain about it. Consider adding all the keys to a new struct and pass that in.

File: src/sources/file.rs
Comment (bruceg) [inline src/sources/file.rs:813]: Do you have any thoughts on accounting for Windows, where the end-of-line sequence is a two-byte pair of characters?

**Review by bruceg — CHANGES_REQUESTED** [review]
Thanks for this contribution, @sproberts92. This generally looks pretty good. I left a couple of questions below. This is missing the external documentation for the new configuration setting, however. That would need to be put in `website/cue/reference/components/sources/file.cue`.

File: src/sources/file.rs
Comment (sproberts92) [reply src/sources/file.rs:813]: Ahh yes, I did not consider Windows' line endings. I'll think about this a little bit, presumably there's a point before the endings are stripped where we can tell.

If I can't find a good way we can fall back to just using the end-of-line offset instead, which is fine for the ordering purpose.

**Comment by sproberts92** [comment]
Thanks for the review @bruceg, I should have some time to make these adjustments on the weekend.

File: src/sources/file.rs
Comment (bruceg) [reply src/sources/file.rs:813]: Getting the beginning of line offset accurately might require plumbing it into the file server code in `lib/file-server/src`, possibly returning it with the bytes in `FileWatcher::read_line`.

Comment (sproberts92) [reply src/sources/file.rs:817]: While looking into this, I noticed that we already have `line_delimiter_as_bytes` in scope, and hence can use the length of that in the calculation instead of my initial fixed `-1`. 

I still agree though that getting this out of `FileWatcher::read_line` instead of trying to re-calculate it here is the more correct approach. The value is definitely available there, it's just a matter of making it available here as well. I'm still looking at how to implement this, but as a relative newcomer to rust there's a bit of a learning curve for me here. Up to you I guess whether you consider the implemented approach good enough for now, or would prefer to wait for the `FileWatcher::read_line` approach.

**Comment by jszwedko** [comment]
Hi @sproberts92 ! Is this ready for re-review? Feel free to tag when it is.

**Comment by sproberts92** [comment]
Hi @jszwedko, @bruceg, thanks for your patience - it can be hard to find the time to work on this in the evenings. This is ready for you to take another look now.

I've made a change to record the line's byte offset at the time it is read instead of trying to calculate it later. Hopefully this is along the lines of what you had in mind @bruceg.

File: src/sources/file.rs
Comment (bruceg) [inline src/sources/file.rs:647]: I think these fields are going to be constant for every event, right? I think it would be best to separate them off so they can be set once and just reused.

**Review by bruceg — COMMENTED** [review]
This is getting close. I have just one concern about efficiency that I left below.

File: lib/file-source/src/file_watcher/mod.rs
Comment (sproberts92) [reply lib/file-source/src/file_watcher/mod.rs]: Yes, needed to also restrict `read_line` though.

**Review by bruceg — APPROVED** [review]
This looks good, thanks for the changes @sproberts92. We'll merge this once the tests complete.

**Comment by sproberts92** [comment]
Apologies @bruceg, I did not realise that `cargo clippy` does not include the tests by default. Resolved now.

**Comment by sproberts92** [comment]
At first glance the failures in the soak test don't appear to be related to my changes. Is this something you've encountered before @bruceg @jszwedko ?

**Comment by bruceg** [comment]
Ah, yes, there was a change to master that caused a configuration breakage in the soaks. Merging with master will fix that.

**Comment by sproberts92** [comment]
Have rebased onto current master.

File: src/sources/file.rs
Comment (sproberts92) [inline src/sources/file.rs:662]: Giving a final look over this, I realise the offset almost certainly should not be cast to a string. Any objections to changing to the following before we merge? 
```suggestion
        event.insert(offset_key.as_str(), offset);
```

Comment (sproberts92) [reply src/sources/file.rs]: In the interest of getting this merged sooner, I went ahead and made the change. This now results in field being serialised to json correctly:
```
{... "offset":248692, ...}
```
instead of
```
{... "offset":"248692", ...}
```

**Comment by jszwedko** [comment]
Thanks @sproberts92 ! I agree with your late assessment that the offset should be an integer. Thanks for making that change.

**Pull request merged**
